### ISSUE TITLE
Metrics (precision, recall, F1-Score)

### ISSUE THREAD

**Issue by anasvaf** [OdysseasKr/neural-disaggregator#10]
Hello Odyssea,

When I run the GRUWithWindow code (`python3 experiment.py kettle`) from the https://github.com/OdysseasKr/online-nilm repository, I get the following numbers for metrics after training the class kettle for 4 houses and testing in the last house.

```
Epoch 00001: saving model to experiments/kettle/CHECKPOINT-kettle-1epochs.hdf5
Epoch 2/3
800000/800000 [==============================] - 722s 902us/step - loss: 0.0039

Epoch 00002: saving model to experiments/kettle/CHECKPOINT-kettle-2epochs.hdf5
Epoch 3/3
800000/800000 [==============================] - 698s 873us/step - loss: 0.0039

Epoch 00003: saving model to experiments/kettle/CHECKPOINT-kettle-3epochs.hdf5
/home/anasvaf/Desktop/MultiLabel_Energy/GRUWithWindow/metrics.py:40: RuntimeWarning: invalid value encountered in true_divide
  return tp/float(tp+fp)
============ Recall: 0.0
============ Precision: nan
============ Accuracy: 0.9962941517081645
============ F1 Score: nan
============ Relative error in total energy: 0.012862401778851902
============ Mean absolute error(in Watts): 21.46095589801764
```

I can confirm also the same exact numbers when I run the RNN code using the https://github.com/OdysseasKr/neural-disaggregator repository.

I am using Python 3.5 and the following library versions:
numpy: 1.15.2
matplotlib: 3.0.0
pandas: 0.23.4
pytables: 3.4.4
nilmtk: 0.2
keras: 2.2.4

Could you help me with that?

Thanks in advance!

-Tasos

**Comment by anasvaf** [comment]
I will do that. Thanks a lot for all the help!

**Comment by maechler** [comment]
I am getting a similar result with `DAEDisaggregator` when I trained `dish washer` on building `1-5` of the `REDD` dataset and then try to evaluate on building `6`.

```
============ Recall: 0.0
============ Precision: nan
============ Accuracy: 0.2995691297390533
============ F1 Score: nan
============ Relative error in total energy: 0.837433614267287
============ Mean absolute error(in Watts): 0.45369448176799543
```

![image](https://user-images.githubusercontent.com/4113819/51255054-bbf06100-19a2-11e9-80d3-4a3144c03e10.png)

When I only train on one building and evaluate it, I do not get the `NaN` errors but still it does not predict anything as my recall stays around `0`.

@anasvaf Could you solve this?
@OdysseasKr Do you think that is a programming error? How much data would be needed to get decent results?

**Comment by maechler** [comment]
@OdysseasKr Thank you very much for your answer! As I had a closer look at the diagram I realised that it is also very strange that the orange graph (ground truth) has its maximum at `4W`. This is not the case when I only train on one building:

![image](https://user-images.githubusercontent.com/4113819/51704535-981fc180-2019-11e9-9647-23ca082441f0.png)

Have you been able to successfully use the `train_across_buildings` method? 

I tried some of the `ukdale-test.py` and `redd-test.py` scripts in this repository, but I never got decent results. I think an additional problem could be that the training data is too sparse. The loss is already very low from the beginning, probably because the neural net learned to predict `0` all the time. I think Jack Kelly in his experiments used a form of artificially generated training data that he computed after extracting all the activations of an appliance. Such training data contains many more occurrences of the trained appliance. 

It would be really nice if you could share how you managed to get good results. E.g.
- What dataset did you use?
- What periods did you use for your train / test data?
- Did you train across multiple buildings?
- Which neural net did you use?
- How many epochs did you train?
- What results did you get?

Any help would really be appreciated!

**Comment by maechler** [comment]
Thanks very much for your response! I understand, but for me training on only one building does not work well either. 

> Are you referring to some results in particular?

I am looking for the optimal parameters to let the example scripts `redd-test.py` or `ukdale-test.py` run and produce good results. When I let them run as they are in this repository, I always get results as mentioned above.

Thanks, I will have a closer look at https://github.com/OdysseasKr/online-nilm, as it seems pretty much configured except for the number of epochs.

**Comment by maechler** [comment]
For anyone interested in this, I managed to get some non constant estimations with `GRUWithWindow` (https://github.com/OdysseasKr/online-nilm) after only 10 epochs of training on the `dish_washer` appliance.

```
============ Recall: 0.3191919191919192
============ Precision: 0.6955245781364637
============ Accuracy: 0.9717776491024899
============ F1 Score: 0.43757212093237946
============ Relative error in total energy: 0.09651187438233116
============ Mean absolute error(in Watts): 25.317596060405524
```

![image](https://user-images.githubusercontent.com/4113819/52531830-0227ae80-2d1c-11e9-8790-229e1326835b.png)
